### PR TITLE
Cheaphashes

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

File: bloomfilter/Data/BloomFilter.hs [hunk 1]
```diff
@@ -75,23 +76,21 @@ module Data.BloomFilter
...
 
 import qualified Data.BloomFilter.BitVec64 as V
 
 -- | An immutable Bloom filter, suitable for querying from pure code.
 data Bloom a = B {
-      hashesN :: {-# UNPACK #-} !Int
-    , shift :: {-# UNPACK #-} !Int
-    , mask :: {-# UNPACK #-} !Int
+      hashesN  :: {-# UNPACK #-} !Int
+    , size     :: {-# UNPACK #-} !Int   -- ^ size is multiple of 64
```
Comment (dcoutts) [inline bloomfilter/Data/BloomFilter.hs:93]: Just clarify that the size unit is bits.
```suggestion
    , size     :: {-# UNPACK #-} !Int   -- ^ Size in bits. This is a multiple of 64
```

File: bloomfilter/Data/BloomFilter/Mutable.hs [hunk 2]
```diff
@@ -81,34 +80,25 @@ import Prelude hiding (elem, length, notElem,
...
-                | numBits > maxHash = maxHash
-                | isPowerOfTwo numBits = numBits
-                | otherwise = nextPowerOfTwo numBits
-        numElems = max 2 (twoBits `unsafeShiftR` logBitsInHash)
-        trueBits = numElems `unsafeShiftL` logBitsInHash
-        shft     = logPower2 trueBits
-        msk      = trueBits - 1
-        isPowerOfTwo n = n .&. (n - 1) == 0
-
+new hash numBits = MB hash numBits' `liftM` V.new numBits'
```
Comment (dcoutts) [inline bloomfilter/Data/BloomFilter/Mutable.hs:83]: The haddock docs above still say

> It is always rounded up to the nearest higher power of two

**Review by dcoutts — APPROVED** [review]
This looks great, both the non-power-of-2 and splitting the hash calculation neatly so we don't need intermediate lists.

One potential bug below for 32bit platforms, or a lurking bug if we switch to 64bit hashes.

**Comment by jorisdral** [comment]
The diff of the PR is showing changes from #87 as well, which should go away if the branch is updated

File: bloomfilter/Data/BloomFilter.hs [hunk 3]
```diff
@@ -176,7 +174,7 @@ singleton hash numBits elt = create hash numBits (\mb -> MB.insert mb elt)
 elem :: Hashable a => a -> Bloom a -> Bool
 elem elt ub = all test (hashes ub elt)
   where test idx' =
-          let !idx = fromIntegral idx' .&. mask ub :: Int
+          let !idx = fromIntegral idx' `rem` size ub :: Int
```
[the remark this answers is not in the document]
Comment (phadej) [reply bloomfilter/Data/BloomFilter.hs]: If we want to support over 512MB bloomfilters, than it cannot be `Hash = Word`. It must be `Word64`, and also indexing into `BitVec64` should use `Int64` (or `Word64`).

**Comment by phadej** [comment]
> Nevertheless, it's worth double checking and seeing if you agree with me or not.

I disagree.

**Pull request merged**